csv/generic/shared/src/test/scala/fs2/data/csv/generic/CsvRowDecoderTest.scala

```diff
@@ -90,4 +97,446 @@ object CsvRowDecoderTest extends SimpleIOSuite {
+      Right(TestData("Bob Smith", "80", "description 2")),
+      Left(new CsvException("Headers have size 3 but row has size 2. Both numbers must match! in line 4", None))
+    )
+
+    val stream = Stream
+      .emit(content)
+      .covary[Fallible]
+      .through(
+        lenient.attemptDecodeGivenHeaders[TestData](separator = ',',
+                                                    skipHeaders = false,
+                                                    headers = NonEmptyList.of("name", "age", "description")))
+      .compile
+      .toList
+
+    stream match {
+      case Right(actual) => expect.eql(expected, actual)
+      case Left(x)       => failure(s"Stream failed with value $x")
+    }
+  }
+
+  pureTest("Parser should return all decoder results when using attemptDecodeSkippingHeaders") {
+    val content =
+      """name-should-not-be-used,age-should-not-be-used,description-should-not-be-used
+        |John Doe,47,description 1
+        |Jane Doe,50
+        |Bob Smith,80,description 2
+        |Alice Grey,78
+        |""".stripMargin
+
+    val expected: List[Either[Throwable, TestData]] = List(
+      Right(TestData("John Doe", "47", "description 1")),
+      Left(new DecoderError("unexpect end of row", None)),
+      Right(TestData("Bob Smith", "80", "description 2")),
+      Left(new DecoderError("unexpect end of row", None))
+    )
+
+    val stream = Stream
+      .emit(content)
+      .covary[Fallible]
+      .through(lenient.attemptDecodeSkippingHeaders[TestData](separator = ','))
+      .compile
+      .toList
+
+    stream match {
+      case Right(actual) => expect.eql(expected, actual)
+      case Left(x)       => failure(s"Stream failed with value $x")
+    }
+  }
+
+  pureTest("Parser should return all decoder results when using attemptDecodeWithoutHeaders") {
+    val content =
+      """John Doe,47,description 1
+        |Jane Doe,50
+        |Bob Smith,80,description 2
+        |Alice Grey,78
+        |""".stripMargin
+
+    val expected: List[Either[Throwable, TestData]] = List(
+      Right(TestData("John Doe", "47", "description 1")),
+      Left(new DecoderError("unexpect end of row", None)),
+      Right(TestData("Bob Smith", "80", "description 2")),
+      Left(new DecoderError("unexpect end of row", None))
+    )
+
+    val stream = Stream
+      .emit(content)
+      .covary[Fallible]
+      .through(lenient.attemptDecodeWithoutHeaders[TestData](separator = ','))
+      .compile
+      .toList
+
+    stream match {
+      case Right(actual) => expect.eql(expected, actual)
+      case Left(x)       => failure(s"Stream failed with value $x")
+    }
+  }
+
+  pureTest("decodeUsingHeaders should map data to case class matching the header name to the case class field name") {
+    implicit val decoder: CsvRowDecoder[Test, String] = testDecoder
+    implicit val testEq: Eq[Test] = Eq.fromUniversalEquals
```

Comment:
Instead of defining it several times (differently) in the test cases, let's move it outside to the other new implicits.